packages/react-router-dom/index.tsx

```diff
@@ -976,6 +977,95 @@ export function useFormAction(
   return createPath(path);
 }
 
+let blockerKey = "blocker-singleton";
+
+export function useBlocker(
+  shouldBlock: boolean | (() => boolean) | BlockerFunction
+) {
+  let { router } = useDataRouterContext(DataRouterHook.UseFetcher);
+
+  let blockerFunction = React.useCallback<BlockerFunction>(() => {
+    return typeof shouldBlock === "function"
+      ? shouldBlock() === true
+      : shouldBlock === true;
+  }, [shouldBlock]);
+
+  let blocker = router.getBlocker(blockerKey, blockerFunction);
+
+  // Cleanup on unmount
+  React.useEffect(() => () => router.deleteBlocker(blockerKey), [router]);
+
+  return blocker;
+}
+
+export function usePrompt(
+  message: string | null | false,
+  opts?: { beforeUnload: boolean }
+) {
+  let { beforeUnload } = opts ? opts : { beforeUnload: false };
+  let blockerFunction = React.useCallback<BlockerFunction>(() => {
+    if (!message) {
+      return false;
+    }
+
+    // Here be dragons! This is not bulletproof (at least at the moment).  If
+    // you click the back button again while the global window.confirm prompt
+    // is open, it returns `false` (telling us to "block") but then _also_
+    // processes the back button click!
+
+    // So, consider:
```

Comment:
I've spent a couple of days working on this and the problem ends up being extremely hairy as each browser has different behavior for 'What happens if you trigger another popstate with the back-button when a prompt is already open'
Chrome: closes the prompt and triggers the 'shouldBlock' conditional case (window.confirm is false) BUT if it triggers another navigation that event gets dropped on the floor as there is a navigation from the back-button pending.
Safari: closes the prompt and triggers the 'shouldBlock' conditional case (same as chrome), BUT if it triggers another navigation that event gets queued and handled in addition to the navigation event from the back-button. 
Firefox: stacks the prompts on top of each other and keeps previous prompts suspended, each is handled correctly as the user clicks 'OK' or 'Cancel'

I've been trying to think of some elegant solutions that don't require browser detection, but still spinning wheels at the moment. Browser detection would probably work but it's a bit gross to hard-code things for what is undefined behavior anyway. 

I haven't tried out Edge or Opera - hoping Edge at least is the same as Chrome.

Reply:
> I've been trying to think of some elegant solutions that don't require browser detection, but still spinning wheels at the moment. Browser detection would probably work but it's a bit gross to hard-code things for what is undefined behavior anyway.

Our general thoughts here are that you really shouldn't block navigations at all unless you accept the tradeoff that in certain scenarios the back button will be broken. We've worked through a lot of this as well and we're unfortunately limited in our ability to make this bullet-proof (which is a big reason why we skipped the implementation in v6 altogether).

We came around on this work because for a lot of folks, preventing data loss through blocking is more important than a clean history stack. Until browser APIs give us something better to work with, this is what we've got. Our plan is to document these tradeoffs and suggest alternative UX approaches where possible. 

All that said ... if our implementation can be better, we're open to that! The main issue I've found is that often we go about patching things for one edge case only to trigger another, and we end up in a never-ending game of whack-a-mole.